consul/server_manager/server_manager.go

```diff
@@ -0,0 +1,340 @@
+		copy(newServers, serverCfg.servers)
+		newServers = append(newServers, server)
+		serverCfg.servers = newServers
+	}
+
+	sm.saveServerConfig(serverCfg)
+}
+
+// cycleServers returns a new list of servers that has dequeued the first
+// server and enqueued it at the end of the list.  cycleServers assumes the
+// caller is holding the serverConfigLock.
+func (sc *serverConfig) cycleServer() (servers []*server_details.ServerDetails) {
+	numServers := len(sc.servers)
+	if numServers < 2 {
+		return servers // No action required
+	}
+
+	newServers := make([]*server_details.ServerDetails, 0, numServers)
+	newServers = append(newServers, sc.servers[1:]...)
+	newServers = append(newServers, sc.servers[0])
+	return newServers
+}
+
+// FindServer takes out an internal "read lock" and searches through the list
+// of servers to find a "healthy" server.  If the server is actually
+// unhealthy, we rely on Serf to detect this and remove the node from the
+// server list.  If the server at the front of the list has failed or fails
+// during an RPC call, it is rotated to the end of the list.  If there are no
+// servers available, return nil.
+func (sm *ServerManager) FindServer() *server_details.ServerDetails {
+	serverCfg := sm.getServerConfig()
+	numServers := len(serverCfg.servers)
+	if numServers == 0 {
+		sm.logger.Printf("[WARN] consul: No servers found in the server config")
+		return nil
+	} else {
+		// Return whatever is at the front of the list because it is
+		// assumed to be the oldest in the server list (unless -
+		// hypothetically - the server list was rotated right after a
+		// server was added).
+		return serverCfg.servers[0]
+	}
+}
+
+// getServerConfig is a convenience method which hides the locking semantics
+// of atomic.Value from the caller.
+func (sm *ServerManager) getServerConfig() serverConfig {
+	return sm.serverConfigValue.Load().(serverConfig)
+}
+
+// New is the only way to safely create a new ServerManager struct.
+func New(logger *log.Logger, shutdownCh chan struct{}, clusterInfo ConsulClusterInfo) (sm *ServerManager) {
+	// NOTE(sean@): Can't pass *consul.Client due to an import cycle
+	sm = new(ServerManager)
+	sm.logger = logger
+	sm.clusterInfo = clusterInfo
+	sm.shutdownCh = shutdownCh
+
+	sc := serverConfig{}
+	sc.servers = make([]*server_details.ServerDetails, 0)
+	sm.saveServerConfig(sc)
+	return sm
+}
+
+// NotifyFailedServer marks the passed in server as "failed" by rotating it
+// to the end of the server list.
+func (sm *ServerManager) NotifyFailedServer(server *server_details.ServerDetails) {
+	serverCfg := sm.getServerConfig()
+
+	// If the server being failed is not the first server on the list,
+	// this is a noop.  If, however, the server is failed and first on
+	// the list, acquire the lock, retest, and take the penalty of moving
+	// the server to the end of the list.
+
+	// Use atomic.CAS to emulate a TryLock().
+	if len(serverCfg.servers) > 0 && serverCfg.servers[0] == server &&
```

Comment:
If you check for `> 1` then this will be cheaper for the case where there's only 1 server. Otherwise you'll keep thrashing to swap it but nothing will really happen in `cycleServer()`.